### PR TITLE
std.uni.isPunctuation: significant performance increase for ASCII

### PR DESCRIPTION
**Pull request by JackStouffer**
With the test character `dchar a = '.';` with one million calls:

```
isPunctuation		79 ms, 385 μs, and 4 hnsecs
isPunctuation2		7 ms, 58 μs, and 9 hnsecs
```

### CODE REVIEW

**Comment by eco** [comment]
What's the performance cost for the non-ASCII case?

It seems like we could avoid the branch by using the type system (i.e. specialize for `std.encoding.AsciiString`) though if this just makes the vast majority of punctuation faster (I know nothing about non-English punctuation) it might as well be considered unicode optimization.

**Comment by JackStouffer** [comment]
> What's the performance cost for the non-ASCII case?

11x slower

> It seems like we could avoid the branch by using the type system

How would that work?

> though if this just makes the vast majority of punctuation faster (I know nothing about non-English punctuation) it might as well be considered unicode optimization

Not the vast majority, unicode supports something like 500+ different punctuation symbols.

**Comment by eco** [comment]
I mean, how much does the added branch hurt performance for the non-ASCII case? I imagine it's very tiny; just curious if you had tested.

> How would that work?

Sorry, just kind of wondering aloud when I saw the added branch about how it could be avoided. I forgot this was dealing with dchar, not string. The pull request looks good though @DmitryOlshansky should probably weigh in.

**Comment by JackStouffer** [comment]
> I mean, how much does the added branch hurt performance for the non-ASCII case? I imagine it's very tiny; just curious if you had tested.

I hadn't. `isASCII` is one `if`statement, so if DMD is inlining correctly the performance should be `<= 1ns`.

File: std/uni.d [hunk 1]
```diff
@@ -9136,7 +9138,20 @@ bool isNumber(dchar c)
 @safe pure nothrow @nogc
 bool isPunctuation(dchar c)
 {
-    return punctuationTrie[c];
+    import std.ascii : isASCII;
+
+    // optimization for ascii case
+    if (c.isASCII)
```
Comment (DmitryOlshansky) [inline std/uni.d]: Just use std.ascii isPunctuation?

Comment (JackStouffer) [inline std/uni.d]: Oh yeah...

**Comment by DmitryOlshansky** [comment]
@JackStouffer One last thing to check is please benchmark your latest version with isASCII check as function.

**Comment by JackStouffer** [comment]
@DmitryOlshansky fixed

```
isPunctuation		77 ms, 169 μs, and 3 hnsecs
isPunctuation2		9 ms, 698 μs, and 8 hnsecs
```

**Comment by JackStouffer** [comment]
Tester failure is unrelated.

**Comment by DmitryOlshansky** [comment]
So somehow it is slower, I suggest to just use < 0x80 in such a case.

**Comment by JackStouffer** [comment]
> So somehow it is slower, I suggest to just use < 0x80 in such a case.

That's only because I am testing the numbers with the debug make build, and therefore nothing is inlined, because it's the only thing that I can get my dmd set up to compile.

**Comment by DmitryOlshansky** [comment]
Thing is std.ascii always had problem with inlining... Anyhow you'd better fix your dmd setup

**Comment by JackStouffer** [comment]
@DmitryOlshansky Because I have no way to prove that the call will be inlined, I changed it to the manual check.

**Comment by JackStouffer** [comment]
@DmitryOlshansky ping

File: std/uni.d [hunk 2]
```diff
@@ -9108,10 +9108,12 @@ bool isMark(dchar c)
 @safe pure nothrow @nogc
 bool isNumber(dchar c)
 {
-    // optimization
-    if (c < 0xAA)
+    import std.ascii : isASCII;
+
+    // optimization for ascii case
+    if (c.isASCII)
```
Comment (DmitryOlshansky) [inline std/uni.d]: Let's be consistent and use constants everywhere

**Comment by JackStouffer** [comment]
@DmitryOlshansky fixed

**Comment by wilzbach** [comment]
LGTM now :)

**Comment by DmitryOlshansky** [comment]
Auto-merge toggled on

**Comment by JackStouffer** [comment]
Thanks!

**Pull request merged**